rubert/src/bin/validate.rs

```diff
@@ -0,0 +1,356 @@
+}
+
+/// Tokenizer configurations.
+struct TokenizerConfig {
+    /// Whether to keep the accents on characters.
+    accents: bool,
+    /// Whether to lowercase words.
+    lowercase: bool,
+    /// The number of tokens for truncation/padding.
+    token_size: usize,
+}
+
+/// Source or target model configurations.
+struct ModelConfig {
+    /// The model kind.
+    kind: ModelKind,
+    /// The path to the vocabulary.
+    vocab: &'static str,
+    /// The path to the model.
+    model: &'static str,
+}
+
+/// Ted talks data configurations.
+struct DataConfig<R: RangeBounds<usize>> {
+    /// The path to the talks.
+    talks: &'static str,
+    /// The range of talks to use for validation.
+    range: R,
+}
+
+/// Combined validation configurations.
+struct ValidatorConfig<R: RangeBounds<usize>> {
+    tokenizer: TokenizerConfig,
+    source: ModelConfig,
+    target: ModelConfig,
+    data: DataConfig<R>,
+}
+
+impl<R: RangeBounds<usize>> ValidatorConfig<R> {
+    /// Builds a validator from this configuration.
+    fn build(&self) -> Validator {
+        Validator::build(self)
+    }
+}
+
+/// The available MBert model pipelines.
+#[allow(clippy::enum_variant_names, clippy::upper_case_acronyms)]
+enum Pipeline {
+    /// A SMBert or QAMBert model pipeline for the onnx runtime.
+    OnnxMBert {
+        tokenizer: Tokenizer<i64>,
+        environment: Environment,
+        session: Option<Session<'static>>,
+        _pinned: PhantomPinned,
+    },
+    /// A SMBert model pipeline for the tract runtime.
+    TractSMBert(BertPipeline<SMBert, NonePooler>),
+    /// A QAMBert model pipeline for the tract runtime.
+    TractQAMBert(BertPipeline<QAMBert, NonePooler>),
+}
+
+impl Pipeline {
+    /// Builds a pipeline from a tokenizer and model configuration.
+    fn build(tokenizer: &TokenizerConfig, model: &ModelConfig) -> Self {
+        match model.kind {
+            ModelKind::OnnxMBert => {
+                let tokenizer = TokenizerBuilder::from_file(model.vocab)
+                    .unwrap()
+                    .with_normalizer(true, false, tokenizer.accents, tokenizer.lowercase)
+                    .with_model("[UNK]", "##", 100)
+                    .with_post_tokenizer("[CLS]", "[SEP]")
+                    .with_truncation(Truncation::fixed(tokenizer.token_size, 0))
+                    .with_padding(Padding::fixed(tokenizer.token_size, "[PAD]"))
+                    .build()
+                    .unwrap();
+                let environment = Environment::builder().build().unwrap();
+                let mut pipeline = Self::OnnxMBert {
+                    tokenizer,
+                    environment,
+                    session: None,
+                    _pinned: PhantomPinned,
+                };
+                if let Self::OnnxMBert {
+                    ref environment,
+                    ref mut session,
+                    ..
+                } = pipeline
+                {
+                    *session = unsafe {
```

Comment:
Can't we build `session` after `environment` and then create the `OnnxMBert` pipeline?

Reply:
this very much depends on how we structure the whole binary. unfortunately the `Session` contains a `&Environment`which leaves us some options:
- create a self-referential struct which requires unsafe code
- have an environment and session separately in the main function which ruins the abstraction layers
- create a session everytime the model is run which is terribly slow

i get why the `onnxruntime` binds the session to the livetime of the environment, but it is a really cumbersome design. also, i noticed that i forgot to actually pin the environment even though i not unpin, but it accidentally worked probably because the memory wasn't moved. anyways i fixed that, updated the safety comment and had to add some more hackery: we can't pin the session together with the environment, because running it requires a mutable reference, which is again due to the `onnxruntime` :/